### PR TITLE
[mono][interp] Expand bblocks reordering with common patterns to facilitate propagation of values

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR expands bblocks reordering with the following common patterns:
- `branch`: a branch instruction (either MINT_IS_CONDITIONAL_BRANCH or MINT_BR).
- `ldc; branch`: a load instruction followed by a binary conditional branch.
- `ldc; compare; branch`: a load instruction followed by a compare instruction and a unary conditional branch.

For example, the following pattern could be inlined into callers bblocks:
```
BB8:
     ldc.i4.0 [var1 <- nil],
     br.s BB1
BB5:
     ldc.i4.1 [var2 <- nil],
     cgt.i4 [var3 <- var1 var2],
     brfalse.i4.s [nil <- var3] BB2
BB3:
     ldstr
```
This PR reorders the code to look like:
```
BB8:
     ldc.i4.0 [var1 <- nil],
     ldc.i4.1 [var2 <- nil],
     cgt.i4 [var3 <- var1 var2],
     brfalse.i4.s [nil <- var3] BB2
BB11:
     br             [nil <- nil], BB3
```
The code is further optimized by constants folding and superins.

Draft PR should confirm if the runtime works as expected.

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.c [hunk 1]
```diff
@@ -8364,49 +8373,141 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+ * - `ldc; branch`: a load instruction followed by a binary conditional branch.
+ * - `compare; branch`: a compare instruction followed by a unary conditional branch.
+ * - `ldc; compare; branch`: a load instruction followed by a compare instruction and a unary conditional branch.
+ */
+static gboolean
+interp_inline_into_callers(InterpBasicBlock *bb, InterpInst** cond_ins) {
+	InterpInst *first = interp_first_ins (bb);
+	InterpInst *second = first ? interp_next_ins(first) : NULL;
+	InterpInst *third = second ? interp_next_ins(second) : NULL;
+
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8424]: Maybe compiler is smart enough. But you could avoid checking for non null with every single if like:
```
if (!first) return;
...
checking patterns with first instruction
...
if (!second) return;
...
checking patterns with first and second instructions
...
```

File: src/mono/mono/mini/interp/transform.c [hunk 2]
```diff
@@ -8364,49 +8373,141 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+/**
+ * Check if the given basic block has a known pattern for inlining into callers blocks, if so, return a pointer to the conditional branch instruction.
+ *
+ * The known patterns are:
+ * - `branch`: a branch instruction (either MINT_IS_CONDITIONAL_BRANCH or MINT_BR).
+ * - `ldc; branch`: a load instruction followed by a binary conditional branch.
+ * - `compare; branch`: a compare instruction followed by a unary conditional branch.
+ * - `ldc; compare; branch`: a load instruction followed by a compare instruction and a unary conditional branch.
+ */
+static gboolean
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8419]: I think the return value of `cond_ins` is rather unnecessary. It could instead directly return the lookup vars.

File: src/mono/mono/mini/interp/transform.c [hunk 3]
```diff
@@ -8364,49 +8373,141 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+
+					last_ins->opcode = cond_ins->opcode;
+					last_ins->sregs [0] = cond_ins->sregs [0];
+					last_ins->sregs [1] = cond_ins->sregs [1];
 					last_ins->info.target_bb = cond_true_bb;
+					InterpInst *src_ins = interp_prev_ins(cond_ins);
+					InterpInst *dest_ins = interp_prev_ins(last_ins);
+
+					// Copy a pattern into caller bb
+					while (src_ins) {
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8495]: This copying here seems a bit hard to follow and unnatural. We could instead just copy all instructions one by one, from `interp_first_ins (bb)` to the end of the `in_bb` (after we got rid of the ending `MINT_BR`).

File: src/mono/mono/mini/interp/transform.c [hunk 4]
```diff
@@ -8438,16 +8538,16 @@ interp_reorder_bblocks (TransformData *td)
 					i++;
 				}
 			}
-		} else if (first->opcode == MINT_BR) {
```
[the remark this answers is not in the document]
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c:8441]: No need, reverted.

File: src/mono/mono/mini/interp/transform.c (hunk 1 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Good point, updated.

File: src/mono/mono/mini/interp/transform.c (hunk 2 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Good point, updated.

File: src/mono/mono/mini/interp/transform.c (hunk 3 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Updated. When copying instructions,`MINT_LDC_I4` has `ins->data [1]` variable and there is a conditional statement for such cases. Other option is to iterate through `ins->data` but it could make the runtime slower.

File: src/mono/mono/mini/interp/transform.c [hunk 5]
```diff
@@ -8364,49 +8373,130 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+	// pattern `branch`
+	if (MINT_IS_CONDITIONAL_BRANCH(first->opcode)) {
+		return first;
+	}
+	
+	InterpInst *second = interp_next_ins(first);
+	if (!second) 
+		return NULL;
+	// pattern `ldc; unop conditional branch`
+	if (MINT_IS_LDC(first->opcode)
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8430]: This pattern is redundant and it is probably not reached anyway. `ldc` + `brtrue/false` is already resolved to `br.s` or optimized out.

File: src/mono/mono/mini/interp/transform.c [hunk 6]
```diff
@@ -8364,49 +8373,130 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+		&& MINT_IS_UNOP_CONDITIONAL_BRANCH(second->opcode) && first->dreg == second->sregs [0]) {
+		return second;
+	}
+	// pattern `ldc; binop conditional branch`
+	if (MINT_IS_LDC(first->opcode)
+		&& MINT_IS_BINOP_CONDITIONAL_BRANCH (second->opcode) && (first->dreg == second->sregs [0] || first->dreg == second->sregs [1])) {
+		return second;
+	}
+	// pattern `compare; unop conditional branch`
+	if (MINT_IS_COMPARE(first->opcode)
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8440]: Do you have a C# example where this case leads to better codegen ? I think we should stick to cases that we can see in practice, and not overdo it with duplicating code.

File: src/mono/mono/mini/interp/transform.c (hunk 5 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Good point. Removed.

File: src/mono/mono/mini/interp/transform.c (hunk 6 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: I don't have such example. Agree, keep it simple and clean.

File: src/mono/mono/mini/interp/transform.c [hunk 7]
```diff
@@ -8366,49 +8375,117 @@ interp_mark_reachable_bblocks (TransformData *td)
...
+ * Returns TRUE if in_bb defines a variable that is used in
+ * the block containing @cond_ins, FALSE otherwise.
+ */
+
 static gboolean
-interp_prev_ins_defines_var (InterpInst *ins, int var1, int var2)
+interp_prev_block_defines_var (InterpInst *in_bb_ins, InterpInst *cond_ins)
 {
+	InterpInst *prev_ins = interp_prev_ins(cond_ins);
+	int var1 = -1, var2 = -1;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8387]: I feel like having the lookup here is not really the best. It doesn't correctly (strictly speaking) initializes the lookup vars. For example in some of the patterns, we know the value of one of the sregs from the existing bblock (`ldc` + `compare`/`condbr`), in which case we should only look for the unkown var. Also we can set these lookup vars directly in `interp_inline_into_callers`, as we check the patterns, without having these separate checks.

**Comment by BrzVlad** [comment]
as with the other PR there is a lot of missing space before `(`

**Comment by kotlarmilos** [comment]
> as with the other PR there is a lot of missing space before (

Thanks for the heads up. Failures seem to be relevant.

File: src/mono/mono/mini/interp/transform.c (hunk 7 above)
Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Good idea, updated.

File: src/mono/mono/mini/interp/transform.c [hunk 8]
```diff
@@ -8380,35 +8393,93 @@ interp_prev_ins_defines_var (InterpInst *ins, int var1, int var2)
...
-					last_ins->sregs [1] = first->sregs [1];
-					last_ins->info.target_bb = cond_true_bb;
+					// Remove ending MINT_BR
+					interp_clear_ins (last_ins);
+					// Copy all instructions one by one, from interp_first_ins (bb) to the end of the in_bb
+					InterpInst *copy_ins = first;
+					while (copy_ins) {
+						InterpInst *new_ins = interp_insert_ins_bb (td, in_bb, in_bb->last_ins, copy_ins->opcode);
+						new_ins->dreg = copy_ins->dreg;
+						new_ins->sregs[0] = copy_ins->sregs[0];
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8472]: space before `[`

File: src/mono/mono/mini/interp/transform.c [hunk 9]
```diff
@@ -8380,35 +8393,93 @@ interp_prev_ins_defines_var (InterpInst *ins, int var1, int var2)
...
+						if (mono_interp_op_sregs[copy_ins->opcode] > 1)
+							new_ins->sregs[1] = copy_ins->sregs[1];
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8474]: space before `[`

File: src/mono/mono/mini/interp/transform.c [hunk 10]
```diff
@@ -8441,7 +8511,7 @@ interp_reorder_bblocks (TransformData *td)
 				}
 			}
 		} else if (first->opcode == MINT_BR) {
-			// All bblocks jumping into this bblock can jump directly into the br target
+			// All bblocks jumping into this bblock can jump directly into the br target only if it is a single instruction in a bb
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/transform.c:8514]: The comment addition seems confusing. Did you mean to say `since it is the single instruction of the bb`

Comment (kotlarmilos) [reply src/mono/mono/mini/interp/transform.c]: Updated. It was confusing and left over from previous changes.

**Review by BrzVlad — APPROVED** [review]
